src/Http/Authentication.Abstractions/src/IAuthenticationResultFeature.cs

```diff
@@ -0,0 +1,16 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace Microsoft.AspNetCore.Authentication
+{
+    /// <summary>
+    /// Used to store the AuthenticateResult for the HttpContext.
+    /// </summary>
+    public interface IAuthenticationResultFeature
+    {
+        /// <summary>
+        /// The <see cref="AuthenticateResult"/> for the request.
+        /// </summary>
+        AuthenticateResult Result { get; set; }
```

Comment:
@davidfowl the AuthenticateResult is partially redundant with User and creates a Source-of-truth/consistency problem. What ensures User and the AuthResult stay in sync? Nothing.

There's a few ways to address this:
A) Do we actually need the whole AuthResult? Maybe we only need to add the AuthProperties and/or error? Those could be added to the existing IHttpAuthenticationFeature using default interface properties.
B) Obsolete IHttpAuthenticationFeature.User in favor of IAuthenticationResultFeature. Have HttpContext.User use the new feature.
C) Have IAuthenticationResultFeature derive from IHttpAuthenticationFeature and have User defer to Result.Principal.

I suggest A.

Reply:
A) sounds good. Lets add the AuthenticationProperties, Ticket, Error, what else?

Reply:
This is two separate things though isn't it?  HttpContext.User is mutable, while this is effectively just exposing the result of AuthenticateAsync(DefaultScheme) as it occurred from the Auth middleware.  It's effectively read only and not a mechanism for modifying the effective user for the request.

Reply:
Hence why I was suggesting we name this something more like IDefaultAuthenticateResultFeature

Reply:
Well the result is able to represent None(no auth info) and Failed(error), we could always just store the User inside the result object if we want it to be consistent, so context.User is setting the underlying results object's user?

Reply:
It's super weird that we'll have two effective users on the request. Developers will not know what to do with that.

Ticket is redundant also, it's just Principal + Properties.

Reply:
@HaoK thoughts?

Reply:
Well I'm just saying something like this:
Context.User // just alias for Context.Authentication.Result.Ticket.Principal
Context.AuthenticateResult // lets you get access to the full result
So then they can also check `Context.AuthenticateResult.None`/`Context.AuthenticateResult.Failure`

Reply:
I don't think we should have AuthenticateResult as a first class property on the HttpContext